providers/digitalocean/digitaloceanProvider.go

```diff
@@ -0,0 +1,176 @@
+}
+
+func (api *DoApi) EnsureDomain(domain string) error {
+	ctx := context.Background()
+	_, resp, err := api.client.Domains.Get(ctx, domain)
+	if resp.Status == "404" {
+		_, _, err := api.client.Domains.Create(ctx, &godo.DomainCreateRequest{
+			Name:      domain,
+			IPAddress: "",
+		})
+		return err
+	} else {
+		return err
+	}
+}
+
+func (api *DoApi) GetNameservers(domain string) ([]*models.Nameserver, error) {
+	return models.StringsToNameservers(defaultNameServerNames), nil
+}
+
+func (api *DoApi) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	ctx := context.Background()
+	dc.Punycode()
+
+	records, _, err := api.client.Domains.Records(ctx, dc.Name, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	existingRecords := make([]*models.RecordConfig, len(records))
+	for i := range records {
+		existingRecords[i] = toRc(dc, &records[i])
+	}
+
+	differ := diff.New(dc)
+	_, create, delete, modify := differ.IncrementalDiff(existingRecords)
+
+	var corrections = []*models.Correction{}
+
+	// Deletes first so changing type works etc.
+	for _, m := range delete {
+		id := m.Existing.Original.(*godo.DomainRecord).ID
+		corr := &models.Correction{
+			Msg: fmt.Sprintf("%s, DO ID: %d", m.String(), id),
+			F: func() error {
+				_, err := api.client.Domains.DeleteRecord(ctx, dc.Name, id)
+				return err
+			},
+		}
+		corrections = append(corrections, corr)
+	}
+	for _, m := range create {
+		req := toReq(dc, m.Desired)
+		corr := &models.Correction{
+			Msg: m.String(),
+			F: func() error {
+				_, _, err := api.client.Domains.CreateRecord(ctx, dc.Name, req)
+				return err
+			},
+		}
+		corrections = append(corrections, corr)
+	}
+	for _, m := range modify {
+		id := m.Existing.Original.(*godo.DomainRecord).ID
+		req := toReq(dc, m.Desired)
+		corr := &models.Correction{
+			Msg: fmt.Sprintf("%s, DO ID: %d", m.String(), id),
+			F: func() error {
+				_, _, err := api.client.Domains.EditRecord(ctx, dc.Name, id, req)
+				return err
+			},
+		}
+		corrections = append(corrections, corr)
+	}
+
+	return corrections, nil
+}
+
+func toRc(dc *models.DomainConfig, r *godo.DomainRecord) *models.RecordConfig {
+	// This handles "@" etc.
+	name := dnsutil.AddOrigin(r.Name, dc.Name)
+
+	target := r.Data
+	// Make target FQDN
+	if r.Type == "CNAME" || r.Type == "MX" || r.Type == "NS" {
+		target = dnsutil.AddOrigin(target+".", dc.Name)
+	}
+
+	return &models.RecordConfig{
+		NameFQDN:     name,
+		Type:         r.Type,
+		Target:       target,
+		TTL:          uint32(r.TTL),
+		MxPreference: uint16(r.Priority),
+		SrvPriority:  uint16(r.Priority),
+		SrvWeight:    uint16(r.Weight),
+		SrvPort:      uint16(r.Port),
+		Original:     r,
+	}
+}
+
+func toReq(dc *models.DomainConfig, rc *models.RecordConfig) *godo.DomainRecordEditRequest {
+	// DO wants the short name, e.g. @
+	name := dnsutil.TrimDomainName(rc.NameFQDN, dc.Name)
+
+	// DO uses the same property for MX and SRV priority
+	priority := 0
+	switch rc.Type {
```

Comment:
Please change:

    switch rc.Type {

to

    switch rc.Type {  // #rtype_variations